lib/utils/qemu.ts

```diff
@@ -79,64 +82,76 @@ export const getQemuPath = function (arch: string) {
 				throw err;
 			})
 			.then(() =>
-				path.join(binDir, `${QEMU_BIN_NAME}-${arch}-${QEMU_VERSION}`),
+				path.join(binDir, `${QEMU_BIN_NAME}-${qemuArch}-${QEMU_VERSION}`),
 			),
 	);
 };
 
-export function installQemu(arch: string) {
-	const request = require('request') as typeof import('request');
-	const fs = require('fs') as typeof import('fs');
-	const zlib = require('zlib') as typeof import('zlib');
-	const tar = require('tar-stream') as typeof import('tar-stream');
-
-	return getQemuPath(arch).then(
-		(qemuPath) =>
-			new Promise(function (resolve, reject) {
-				const installStream = fs.createWriteStream(qemuPath);
-
-				const qemuArch = balenaArchToQemuArch(arch);
-				const fileVersion = QEMU_VERSION.replace('v', '').replace('+', '.');
-				const urlFile = encodeURIComponent(
-					`qemu-${fileVersion}-${qemuArch}.tar.gz`,
-				);
-				const urlVersion = encodeURIComponent(QEMU_VERSION);
-				const qemuUrl = `https://github.com/balena-io/qemu/releases/download/${urlVersion}/${urlFile}`;
-
-				const extract = tar.extract();
-				extract.on('entry', function (header, stream, next) {
+async function installQemu(arch: string, qemuPath: string) {
+	const qemuArch = balenaArchToQemuArch(arch);
+	const fileVersion = QEMU_VERSION.replace('v', '').replace('+', '.');
+	const urlFile = encodeURIComponent(`qemu-${fileVersion}-${qemuArch}.tar.gz`);
+	const urlVersion = encodeURIComponent(QEMU_VERSION);
+	const qemuUrl = `https://github.com/balena-io/qemu/releases/download/${urlVersion}/${urlFile}`;
+
+	const request = await import('request');
+	const fs = await import('fs');
+	const zlib = await import('zlib');
+	const tar = await import('tar-stream');
+
+	// createWriteStream creates a zero-length file on disk that
+	// needs to be deleted if the download fails
+	const installStream = fs.createWriteStream(qemuPath);
+	try {
+		await new Promise<void>((resolve, reject) => {
+			const extract = tar.extract();
+			extract.on('entry', function (header, stream, next) {
+				try {
 					stream.on('end', next);
 					if (header.name.includes(`qemu-${qemuArch}-static`)) {
 						stream.pipe(installStream);
 					} else {
 						stream.resume();
 					}
+				} catch (err) {
+					reject(err);
+				}
+			});
+			request(qemuUrl)
+				.on('error', reject)
+				.pipe(zlib.createGunzip())
+				.on('error', reject)
+				.pipe(extract)
+				.on('error', reject)
+				.on('finish', function () {
+					fs.chmodSync(qemuPath, '755');
+					resolve();
 				});
-
-				return request(qemuUrl)
-					.on('error', reject)
-					.pipe(zlib.createGunzip())
-					.on('error', reject)
-					.pipe(extract)
-					.on('error', reject)
-					.on('finish', function () {
-						fs.chmodSync(qemuPath, '755');
-						resolve();
-					});
-			}),
-	);
+		});
+	} catch (err) {
+		try {
+			await fs.promises.unlink(qemuPath);
```

Comment:
The `fs.createWriteStream(qemuPath)` call earlier in this function creates a zero-length file on disk. On successful download, the file is expanded to the size of the download. On errors, before this PR the file was left behind and, the next time the CLI was executed, the CLI would assume the file no longer needed to be downloaded (a bug). This `unlink` call deletes "corrupt" files.